### PR TITLE
Emit timestamps on each line of output

### PR DESCRIPTION
**Pull request by karkhaz**
This commit adds the --timestamp command line option to allow cprover
tools to emit timestamps before each line of output. The timestamps are
either pretty-formatted wall-clock times (which may decrease as time
passes in the event of a leap-second or daylight savings time event), or
raw numbers that are guaranteed to increase. In either case, the
timestamps have microsecond precision if supported by the underlying
platform.

### CODE REVIEW

File: src/util/timestamper.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    return message;
+  }
+  static const timestampert *const make(clockt clock_type);
+};
+
+
+class monotonic_timestampert : public timestampert
+{
+public:
+  virtual const std::string stamp(const std::string &message) const override;
```
Comment (smowton) [inline src/util/timestamper.h]: I believe `const` on a by-value return type doesn't really mean anything

File: src/util/timestamper.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+#include "invariant.h"
+
+#define OPT_TIMESTAMP "(timestamp):"
+
+#define HELP_TIMESTAMP \
+" --timestamp <monotonic|wall> print microsecond-precision timestamps.\n" \
+"                              monotonic: timestamps are relative to an\n" \
+"                              arbitrary start point, guaranteed to " \
+"increase,\n" \
```
Comment (smowton) [inline src/util/timestamper.h]: fix this formatting

File: src/util/ui_message.cpp [hunk 3]
```diff
@@ -76,6 +87,7 @@ ui_message_handlert::~ui_message_handlert()
   case uit::PLAIN:
     break;
   }
+  delete time;
```
Comment (smowton) [inline src/util/ui_message.cpp]: Use a `unique_ptr` rather than hand-pairing `new` and `delete`

**Review by smowton — COMMENTED** [review]
Seems reasonable; some suggestions, and I suggest rather than have a `stamp` method that takes a message and returns it with something prepended, just provide a `current_time_string` method that returns the current time and have `messaget` do the concatenation.

File: src/cbmc/cbmc_parse_options.h [hunk 4]
```diff
@@ -70,7 +72,7 @@ class optionst;
   OPT_GOTO_TRACE \
   "(fixedbv)(floatbv)(all-claims)(all-properties)" // legacy, and will eventually disappear // NOLINT(whitespace/line_length)
 
-class cbmc_parse_optionst:
+class cbmc_parse_optionst :
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.h]: Please avoid pure whitespace changes.

File: src/goto-instrument/goto_instrument_parse_options.h [hunk 5]
```diff
@@ -82,7 +84,7 @@ Author: Daniel Kroening, kroening@kroening.com
   "(splice-call):" \
 
 
-class goto_instrument_parse_optionst:
+class goto_instrument_parse_optionst :
```
Comment (martin-cs) [inline src/goto-instrument/goto_instrument_parse_options.h]: Ditto.

**Review by martin-cs — COMMENTED** [review]
I can see how this would be useful but I'm a little concerned about the duplication of functionality between timestamper.h and our existing time.h.  Doesn't this give two different timing APIs?

**Comment by karkhaz** [comment]
@smowton thanks, I've applied your suggestions.
@martin-cs thanks, I've applied your suggestions. I propose that we remove time.h and time_stopping.h entirely, since these were written in 2004 before the C++11 time API existed. I've updated the patchset to remove the old time API and make all clients of the old API use C++11 std::chrono. Thoughts?

File: src/util/timestamper.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+#include <memory>
+#include <sstream>
+
+/// \brief Timestamp class hierarchy
+///
+/// This class hierarchy supports generation of timestamps in various
+/// textual formats. The timestamps returned by instances of this class
+/// are empty; more meaningful timestamps are returned by derived classes.
+class timestampert
```
Comment (smowton) [inline src/util/timestamper.h]: Clarify whether this abstract, or is instantiating the base useful / sensible?

Comment (karkhaz) [reply src/util/timestamper.h]: This is not abstract, it is instantiated when `--timestamps` is not passed on the command line. This means that ui_messaget can always call `time->stamp()` without bothering to check whether `--timestamps` was passed; virtual dispatch takes care of whether to emit an actual timestamp or just the empty string. Should I clarify this in comments?

Comment (smowton) [reply src/util/timestamper.h]: Yes please, it looks at the moment like an abstract base that provides a default implementation for a  concrete subclass

**Review by smowton — APPROVED** [review]
LGTM

File: src/cbmc/all_properties.cpp [hunk 7]
```diff
@@ -56,8 +57,9 @@ safety_checkert::resultt bmc_all_propertiest::operator()()
 
   solver.set_message_handler(get_message_handler());
 
-  // stop the time
-  absolute_timet sat_start=current_time();
+  std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>
+    sat_start = std::chrono::time_point_cast<std::chrono::seconds>(
+      std::chrono::steady_clock::now());
```
Comment (tautschnig) [inline src/cbmc/all_properties.cpp]: Would it be worthwhile keeping a `time_stopping.h` (but really just the header) with

````c++
#include <chrono>

using absolute_timet = 
  std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>;

absolute_timet current_time()
{
  return
    std::chrono::time_point_cast<std::chrono::seconds(
      std::chrono::steady_clock::now());
}
````

File: .gitignore [hunk 8]
```diff
@@ -45,7 +45,7 @@ src/ansi-c/gcc_builtin_headers_tm.inc
 src/ansi-c/gcc_builtin_headers_mips.inc
 src/ansi-c/gcc_builtin_headers_power.inc
 src/ansi-c/gcc_builtin_headers_ubsan.inc
-src/java_bytecode/java_core_models.inc
+src/java_bytecode/library/java_core_models.inc
```
Comment (tautschnig) [inline .gitignore]: For future reference: please put those changes in a separate PR as it is trivial to merge (and would avoid others accidentally coming up with the same).

File: src/cbmc/all_properties.cpp [hunk 9]
```diff
@@ -57,9 +57,10 @@ safety_checkert::resultt bmc_all_propertiest::operator()()
 
   solver.set_message_handler(get_message_handler());
 
+  std::chrono::steady_clock::time_point now =
+    std::chrono::steady_clock::now();
   std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>
-    sat_start = std::chrono::time_point_cast<std::chrono::seconds>(
-      std::chrono::steady_clock::now());
+    sat_start = std::chrono::time_point_cast<std::chrono::seconds>(now);
```
Comment (tautschnig) [inline src/cbmc/all_properties.cpp]: Is this change (and the one above) possibly in the wrong commit?

File: src/cbmc/cbmc_parse_options.cpp [hunk 10]
```diff
@@ -1029,6 +1029,7 @@ void cbmc_parse_optionst::help()
     " --json-ui                    use JSON-formatted output\n"
     HELP_GOTO_TRACE
     " --verbosity #                verbosity level\n"
+    HELP_TIMESTAMP
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp:1032]: Please add to goto-analyzer, goto-diff as well.

File: src/util/timestamper.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+Author: Kareem Khazem <karkhaz@karkhaz.com>
+
+\*******************************************************************/
+
+
+#include <chrono>
+#include <cstdlib>
+
+#include "invariant.h"
+#include "timestamper.h"
```
Comment (tautschnig) [inline src/util/timestamper.cpp]: Nit pick: I think we've move to including the header file associated with an implementation to the top of the list of includes.

File: src/util/timestamper.h [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+"                              wall: timestamps use the format\n"          \
+"                              %Y-%m-%dT%H:%M:%S.u where 'u' is a six-\n"  \
+"                              digit zero-padded microsecond counter.\n"   \
+"                              Timestamp may decrease on daylight saving\n"\
+"                              time changes\n"
+#endif
+
+
+#include <memory>
+#include <sstream>
```
Comment (tautschnig) [inline src/util/timestamper.h]: Where is this used/required?

File: src/util/timestamper.h [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  enum class clockt
+  {
+    /// \ref timestampert
+    NONE,
+    /// monotonic_timestampert
+    MONOTONIC,
+    /// wall_clock_timestampert
+    WALL_CLOCK
+  };
+  timestampert() {}
```
Comment (tautschnig) [inline src/util/timestamper.h]: Seems unnecessary as it is implicit?

File: src/util/timestamper.h [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  virtual ~timestampert() {}
```
Comment (tautschnig) [inline src/util/timestamper.h]: I'd suggest `virtual ~timestampert() = default;`

File: src/util/timestamper.h [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+  /// \brief Default timestamp: the empty string
+  virtual std::string stamp() const
```
Comment (tautschnig) [inline src/util/timestamper.h]: It seems you'd need `#include <string>` rather than `<sstream>`?

File: src/util/timestamper.h [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+class wall_clock_timestampert : public timestampert
+{
+public:
+  /// \brief See \ref HELP_TIMESTAMP in util/timestamper.h for time
+  /// stamp format
+  virtual std::string stamp() const override;
+};
+
+#endif
+#endif
```
Comment (tautschnig) [inline src/util/timestamper.h]: Nit picks: add `/* CPROVER_UTIL_TIMESTAMPER_H */` in the very last line, and swap the blank line and the `#endif` before.

File: src/util/timestamper.cpp [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+std::string monotonic_timestampert::stamp() const
+{
+  std::chrono::time_point<std::chrono::steady_clock, std::chrono::microseconds>
+    time_stamp = std::chrono::time_point_cast<std::chrono::microseconds>(
+      std::chrono::steady_clock::now());
+
+  auto cnt = time_stamp.time_since_epoch().count();
+  std::div_t divmod = div(cnt, 1000000);
+  char u_seconds_fmt[7];
+  std::snprintf(u_seconds_fmt, sizeof(u_seconds_fmt), "%06d", divmod.rem);
```
Comment (tautschnig) [inline src/util/timestamper.cpp]: How about `std::setfill('0') << std::setw(6) << divmod.rem` to get rid of `snprintf`?

File: src/util/timestamper.cpp [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+  unsigned u_seconds = time_stamp.time_since_epoch().count() % 1000000;
+  char u_seconds_fmt[7];
+  std::snprintf(u_seconds_fmt, sizeof(u_seconds_fmt), "%06u", u_seconds);
+
+  std::time_t tt = std::chrono::system_clock::to_time_t(time_stamp);
+  std::tm *local = std::localtime(&tt);
+  char time_buf[WALL_TEXT_BUFFER_LENGTH];
+  int chars =
+    std::strftime(time_buf, WALL_TEXT_BUFFER_LENGTH, WALL_FORMAT, local);
```
Comment (tautschnig) [inline src/util/timestamper.cpp]: You might use `std::put_time` instead?

File: src/util/ui_message.cpp [hunk 19]
```diff
@@ -57,7 +61,19 @@ ui_message_handlert::ui_message_handlert(
...
+    cmdline.isset("timestamp")
+      ? cmdline.get_value("timestamp") == "monotonic"
+          ? timestampert::clockt::MONOTONIC
+          : cmdline.get_value("timestamp") == "wall"
+              ? timestampert::clockt::WALL_CLOCK
+              : timestampert::clockt::NONE
+      : timestampert::clockt::NONE
+#else
+      timestampert::clockt::NONE
+#endif
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: I don't think you need the `#ifndef _WIN32` here: for `_WIN32`, your code simply ignores the `clock_type` argument, and hence you're completely safe to us the full ternary branch above in either case!?

**Review by tautschnig — COMMENTED** [review]
No blockers, but a few ideas/suggestions.

**Comment by karkhaz** [comment]
Thanks @tautschnig, I've applied all your suggestions. The code is also now (finally) building on Windows. I've also fixed up everything that clang-format was complaining about, apart from the usual noise from the parse_options strings.

**Review by tautschnig — APPROVED** [review]

File: .gitignore (hunk 8 above)
Comment (karkhaz) [reply .gitignore]: Acknowledged, will do this next time!

File: src/cbmc/cbmc_parse_options.cpp (hunk 10 above)
Comment (martin-cs) [reply src/cbmc/cbmc_parse_options.cpp:1032]: +1 that would be lovely.

File: src/util/timestamper.h [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+  "                              arbitrary start point, guaranteed to\n"       \
+  "                              increase, formatted as s.u where 's'\n"       \
+  "                              increases every second and 'u' is a zero-\n"  \
+  "                              padded six-digit microsecond counter.\n"      \
+  "                              wall: timestamps use the format\n"            \
+  "                              %Y-%m-%dT%H:%M:%S.u where 'u' is a six-\n"    \
+  "                              digit zero-padded microsecond counter.\n"     \
+  "                              Timestamp may decrease on daylight saving\n"  \
+  "                              time changes\n"
+#endif
```
Comment (martin-cs) [inline src/util/timestamper.h]: Help messages are good.  Verbose help messages are good.  Help messages that are significantly more verbose than all of their friends ... might be better as a slightly more succinct help message and an entry in the manual.

**Review by martin-cs — APPROVED** [review]
Comments but no show stoppers.

**Comment by martin-cs** [comment]
@karkhaz : as regards API changes.  For me having one, good, consistent API is the key thing.  Your approach to fixing it sounds reasonable.  Certainly using std functions avoids the usual cross-platform mess that is time keeping.

File: src/cbmc/cbmc_parse_options.cpp (hunk 10 above)
Comment (karkhaz) [reply src/cbmc/cbmc_parse_options.cpp:1032]: This is done

**Comment by karkhaz** [comment]
@martin-cs not sure what you're +1ing on `cbmc_parse_options:1032`, is it @tautschnig's suggestion that I add this option to goto_{analyzer,diff} also? If so that's done.

I've just rebased and made the help text more succinct. Thank you!

**Comment by martin-cs** [comment]
@karkhaz It was and yes, done.  Thanks.  This is OK by me once CI passes.

**Comment by tautschnig** [comment]
@karkhaz It seems time_stopping.cpp and timer.cpp have been removed from the build, but not actually from the repository?

**Comment by karkhaz** [comment]
@tautschnig Ah, thanks! I accidentally destroyed the commit that replaced time{r,_stopping} with std::chrono. I've just re-written it, but I'm afraid I'm working blindly: the build seems to be broken on my machine, I'll see what CI says.

**Comment by karkhaz** [comment]
(nice thing about inverted sleep-schedule is that I have the CI server to myself)

**Comment by karkhaz** [comment]
@tautschnig I've updated the PR to remove the old time API like before. For some reason, it's failing some of its tests, I'm not sure why---may I have some guidance on how to debug this?

**Pull request merged**